File: src/terminal.c
/*
Written by Justus Wolff in very late 2025.
*/
#include <stdlib.h>
#include "terminal.h"
#include "common.h"
#include "font8x8_basic.h"

/* color table
0 	#000000 	black
1 	#AA0000 	dark red
2 	#00AA00 	dark green
3 	#AAAA00 	dark yellow
4 	#0000AA 	dark blue
5 	#AA00AA 	dark magenta
6 	#00AAAA 	dark cyan
7 	#AAAAAA 	light grey
8 	#555555 	dark grey
9 	#FF5555 	light red
10 	#55FF55 	light green
11 	#FFFF55 	light yellow
12 	#5555FF 	light blue
13 	#FF55FF 	light magenta
14 	#55FFFF 	light cyan
15 	#FFFFFF 	white
*/

typedef uint8_t VM_pixel;
VM_pixel VM_colortable[16][3] = {
    {0x00, 0x00, 0x00}, // black
    {0x00, 0x00, 0xAA}, // dark blue
    {0x00, 0xAA, 0x00}, // dark green
    {0x00, 0xAA, 0xAA}, // dark cyan
    {0xAA, 0x00, 0x00}, // dark red
    {0xAA, 0x00, 0xAA}, // dark magenta
    {0xAA, 0xAA, 0x00}, // dark yellow
    {0xAA, 0xAA, 0xAA}, // light grey
    {0x55, 0x55, 0x55}, // dark grey
    {0x55, 0x55, 0xFF}, // light blue
    {0x55, 0xFF, 0x55}, // light green
    {0x55, 0xFF, 0xFF}, // light cyan
    {0xFF, 0x55, 0x55}, // light red
    {0xFF, 0x55, 0xFF}, // light magenta
    {0xFF, 0xFF, 0x55}, // light yellow
    {0xFF, 0xFF, 0xFF}, // white
};


VM_term VM_newterm(uint8_t charsnh, uint8_t charsnv) {
    VM_term out;
    out.charsnh = charsnh;
    out.charsnv = charsnv;
    uint32_t pixam = (8 * charsnh) * (8 * charsnv);
    out.pixbuf = (uint8_t*)calloc(pixam, sizeof(uint8_t));
    out.nlchar = '\n';
    return out;
}
void VM_delterm(VM_term* term) {
    free(term->pixbuf);
    term->pixbuf = NULL;
}
void VM_setrawpix(VM_term* term, uint32_t x, uint32_t y, SDL_Renderer* renderer, uint8_t color) {
    term->pixbuf[x+(y*term->charsnh*8)] = color;
	//SDL_SetRenderDrawColor(renderer, color[0], color[1], color[2], 255);
    //SDL_RenderDrawPoint(renderer, x, y);
}
void VM_setpix(VM_term* term, uint32_t x, uint32_t y, uint8_t colorindex, SDL_Renderer* renderer) {
    //uint8_t r,g,b;
    //r = VM_colortable[colorindex][0];
    //g = VM_colortable[colorindex][1];
    //b = VM_colortable[colorindex][2];
    VM_setrawpix(term, x, y, renderer, colorindex);
}
void VM_setchar(VM_term* term, uint8_t fcolor, uint8_t bcolor, uint8_t charindex, uint8_t column, uint8_t row, SDL_Renderer* renderer) {
    if (charindex > 127) {
        charindex = 0x00;
    }
    for (uint8_t y=0;y<8;y++) {
        for (uint8_t x=0;x<8;x++) {
            VM_setpix(term, x+(column*8), y+(row*8), font8x8_basic[charindex][y]>>x & 1 ? fcolor : bcolor, renderer);
        }
    }
}
void VM_copypix(VM_term* term, uint32_t sx, uint32_t sy, uint32_t dx, uint32_t dy, SDL_Renderer* renderer) {
    VM_setpix(term, dx, dy, term->pixbuf[sx+(sy*(8*term->charsnh))], renderer);
}
void VM_copycharpix(VM_term* term, uint8_t sx, uint8_t sy, uint8_t dx, uint8_t dy, SDL_Renderer* renderer) {
    sy *= 8;
    sx *= 8;
    dx *= 8;
    dy *= 8;
    for (uint8_t y=0;y<8;y++) {
        for (uint8_t x=0;x<8;x++) {
            VM_copypix(term, sx+x, sy+y, dx+x, dy+y, renderer);
        }
    }
}

File: src/terminal.rs
use crate::font::FONT_8X8;

/// 16-color palette (CGA-style ordering).
pub const COLOR_TABLE: [[u8; 3]; 16] = [
    [0x00, 0x00, 0x00], // 0  black
    [0x00, 0x00, 0xAA], // 1  dark blue
    [0x00, 0xAA, 0x00], // 2  dark green
    [0x00, 0xAA, 0xAA], // 3  dark cyan
    [0xAA, 0x00, 0x00], // 4  dark red
    [0xAA, 0x00, 0xAA], // 5  dark magenta
    [0xAA, 0xAA, 0x00], // 6  dark yellow
    [0xAA, 0xAA, 0xAA], // 7  light grey
    [0x55, 0x55, 0x55], // 8  dark grey
    [0x55, 0x55, 0xFF], // 9  light blue
    [0x55, 0xFF, 0x55], // 10 light green
    [0x55, 0xFF, 0xFF], // 11 light cyan
    [0xFF, 0x55, 0x55], // 12 light red
    [0xFF, 0x55, 0xFF], // 13 light magenta
    [0xFF, 0xFF, 0x55], // 14 light yellow
    [0xFF, 0xFF, 0xFF], // 15 white
];

pub struct Terminal {
    pub charsnh: u8,
    pub charsnv: u8,
    /// Pixel buffer: one byte per pixel (color index), row-major.
    /// Size = (8 * charsnh) * (8 * charsnv).
    pub pixbuf: Vec<u8>,
    pub colors: u8,
    pub hrange: u16,
    pub vrange: u16,
    pub cursor: u16,
    pub nlchar: u8,
    pub scrollmask: u32,
    pub char0even: u32,
    pub char0odd: u32,
}

impl Terminal {
    pub fn new(charsnh: u8, charsnv: u8) -> Self {
        let pix_count = (8 * charsnh as usize) * (8 * charsnv as usize);
        Terminal {
            charsnh,
            charsnv,
            pixbuf: vec![0; pix_count],
            colors: 0,
            hrange: 0,
            vrange: 0,
            cursor: 0,
            nlchar: b'\n',
            scrollmask: 0,
            char0even: 0,
            char0odd: 0,
        }
    }

    pub fn pixel_width(&self) -> u32 {
        8 * self.charsnh as u32
    }

    pub fn pixel_height(&self) -> u32 {
        8 * self.charsnv as u32
    }

    fn set_pix(&mut self, x: u32, y: u32, color_index: u8) {
        let stride = 8 * self.charsnh as u32;
        let idx = x + y * stride;
        if (idx as usize) < self.pixbuf.len() {
            self.pixbuf[idx as usize] = color_index;
        }
    }

    fn get_pix(&self, x: u32, y: u32) -> u8 {
        let stride = 8 * self.charsnh as u32;
        let idx = x + y * stride;
        if (idx as usize) < self.pixbuf.len() {
            self.pixbuf[idx as usize]
        } else {
            0
        }
    }

    pub fn set_char(
        &mut self,
        fcolor: u8,
        bcolor: u8,
        charindex: u8,
        column: u8,
        row: u8,
    ) {
        let ci = if charindex > 127 { 0 } else { charindex };
        for y in 0u8..8 {
            for x in 0u8..8 {
                let lit = (FONT_8X8[ci as usize][y as usize] >> x) & 1 != 0;
                let color = if lit { fcolor } else { bcolor };
                self.set_pix(
                    x as u32 + column as u32 * 8,
                    y as u32 + row as u32 * 8,
                    color,
                );
            }
        }
    }

    fn copy_char_pix(&mut self, sx: u8, sy: u8, dx: u8, dy: u8) {
        let sx = sx as u32 * 8;
        let sy = sy as u32 * 8;
        let dx = dx as u32 * 8;
        let dy = dy as u32 * 8;
        for y in 0u32..8 {
            for x in 0u32..8 {
                let c = self.get_pix(sx + x, sy + y);
                self.set_pix(dx + x, dy + y, c);
            }
        }
    }

    /// Plot a single pixel via the pixel-plot I/O range.
    /// `offset` = address offset from 0x9FE0 (lower 4 bits = color index).
    pub fn plot_pixel(&mut self, val: u32, offset: u16) {
        let color_index = (offset & 0b1111) as u8;
        let row = ((val >> 8) & 0xFF) as u32;
        let column = (val & 0xFF) as u32;
        self.set_pix(column, row, color_index);
    }

    /// Scroll-print I/O handler.
    /// `val` = value written, `offset` = address offset from I/O base (0..=0x3F).
    /// The offset bits encode control flags.
    pub fn scroll_print(&mut self, val: u32, offset: u16) {
        let nlchar_flag = (offset >> 5) & 1 != 0;
        let tmscroll = (offset >> 4) & 1 != 0;
        // bit 3 unused
        let roprint = (offset >> 2) & 1 != 0;
        let cfdata = (offset >> 1) & 1 != 0;
        let etmode = offset & 1 != 0;

        let mut column = (self.cursor & 0b11111) as u8;
        let mut row = ((self.cursor >> 5) & 0b11111) as u8;

        let (prange, srange) = if roprint {
            (self.hrange, self.vrange)
        } else {
            (self.vrange, self.hrange)
        };

        let pend = ((prange >> 5) & 0b11111) as u8;
        let pstart = (prange & 0b11111) as u8;
        let send = ((srange >> 5) & 0b11111) as u8;
        let sstart = (srange & 0b11111) as u8;

        let forecolor = if !cfdata {
            self.colors & 0b1111
        } else {
            ((val >> 8) & 0b1111) as u8
        };
        let backcolor = if !cfdata {
            (self.colors >> 4) & 0b1111
        } else {
            ((val >> 13) & 0b1111) as u8
        };
        let charindex = (val & 0xFF) as u8;

        // Map primary/secondary to column/row
        let mut pval = if roprint { column } else { row };
        let mut sval = if roprint { row } else { column };

        // Newline character handling
        if nlchar_flag && charindex == self.nlchar {
            pval = pend.wrapping_add(1);
        }

        if etmode {
            // Character print mode
            if pval > pend {
                pval = 0;
                sval = sval.wrapping_add(1);
            }
            if sval > send {
                if tmscroll {
                    // Scroll: copy lines up
                    for y in sstart..=send {
                        for x in pstart..=pend {
                            self.copy_char_pix(x, y.wrapping_add(1), x, y);
                        }
                    }
                    // Fill last line
                    for x in pstart..=pend {
                        self.set_char(forecolor, backcolor, self.nlchar, x, send);
                    }
                    sval = sval.wrapping_sub(1);
                } else {
                    sval = sstart;
                }
            }

            if !(nlchar_flag && charindex == self.nlchar) {
                // Write character at current position
                let (col, rw) = if roprint {
                    (pval, sval)
                } else {
                    (sval, pval)
                };
                self.set_char(forecolor, backcolor, charindex, col, rw);
                pval = pval.wrapping_add(1);
            }
        } else {
            // Scroll-only mode
            for y in sstart..send {
                for x in pstart..=pend {
                    self.copy_char_pix(x, y.wrapping_add(1), x, y);
                }
            }
            // Fill last line with the character
            for x in pstart..=pend {
                self.set_char(forecolor, backcolor, charindex, x, send);
            }
        }

        // Write back cursor
        if roprint {
            column = pval;
            row = sval;
        } else {
            row = pval;
            column = sval;
        }
        self.cursor = column as u16 + ((row as u16) << 5);
    }
}
